src/udev/udevadm-wait.c

```diff
@@ -0,0 +1,298 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+
+#include <getopt.h>
+#include <unistd.h>
+
+#include "sd-event.h"
+
+#include "alloc-util.h"
+#include "chase-symlinks.h"
+#include "errno-util.h"
+#include "fd-util.h"
+#include "inotify-util.h"
+#include "parse-util.h"
+#include "path-util.h"
+#include "static-destruct.h"
+#include "strv.h"
+#include "udev-util.h"
+#include "udevadm.h"
+
+static usec_t arg_timeout_usec = USEC_INFINITY;
+static bool arg_initialized = true;
+static bool arg_removed = false;
+static bool arg_settle = false;
+static char **arg_devices = NULL;
+
+STATIC_DESTRUCTOR_REGISTER(arg_devices, strv_freep);
+
+static int check_device(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        int r;
+
+        assert(path);
+
+        if (access(path, F_OK) < 0)
+                return arg_removed ? true : -errno;
+
+        if (arg_removed)
+                return false;
+
+        if (!arg_initialized)
+                return true;
+
+        r = sd_device_new_from_path(&dev, path);
+        if (r < 0)
+                return r;
+
+        return sd_device_get_is_initialized(dev);
+}
+
+static bool check(void) {
+        int r;
+
+        if (arg_settle) {
+                r = udev_queue_is_empty();
+                if (r <= 0) {
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to check if udev queue is empty, assuming not empty: %m");
+                        return false;
+                }
+        }
+
+        STRV_FOREACH(p, arg_devices) {
+                r = check_device(*p);
+                if (r <= 0) {
+                        if (r < 0 && !ERRNO_IS_DEVICE_ABSENT(r))
+                                log_warning_errno(r, "Failed to check if device \"%s\" %s, assuming %s.: %m",
+                                                  *p,
+                                                  arg_removed ? "is removed" : arg_initialized ? "is initialized" : "exists",
+                                                  arg_removed ? "not removed" : arg_initialized ? "not initialized" : "not exist");
+                        return false;
+                }
+        }
+
+        return true;
+}
+
+static int device_monitor_handler(sd_device_monitor *monitor, sd_device *device, void *userdata) {
```

Comment:
hmm, i think we should probably use the data in sd_device a bit before checking in /sys. i.e. add a helper

```c
static int device_matches_path(sd_device *d);
```

or so that goes through all relevant paths in the supplied device object, i.e. the syspath, the devpaths and the symlinks. And only if at least one of them matches against `arg_devices` we should go for check().

I am a bit afraid that we'll get a ton of events here, and if we never bother with the event data we'll always end up accessing /dev/ and/or /sys/ to verify the state. I think it would be much better if we would instead try to already do superficial verification via the data we got sent anyway, and only if it that looks good, check fully.

(in fact, I'd go even further: extend the `check()` routine, so that it gets an additional, optional `sd_device *focus` parameter or so, that we will initialize from the monitor callback device object. `check()` should skip ny further verification of all device paths that match this object as we *know* the device is already present if we got an event for it.)